### ISSUE TITLE
SslStream.AuthenticateAsServerAsync is significantly slower in CoreFX 4.4

### ISSUE THREAD

**Issue by cesarbs** [dotnet/corefx#14954]
Kestrel perf tests recently started yielding 0 RPS on HTTPS. This happens when we open a large number of simultaneous connections to Kestrel.

What I've discovered so far is that `SslStream.AuthenticateAsServerAsync` is taking significantly longer to run.

Opening 256 connections and measuring the time that takes to run in Kestrel 1.1.0 (which referenced CoreFX 4.3) I get an average of 339 ms per call.

Running against our dev branch (referencing CoreFX 4.4 packages), I'm seeing an average of 10 **seconds** per call, and some of the tasks never appear to complete in a reasonable time.

The delay appears to be proportional to the number of simultaneous calls we're making. Here are some rough measurements of the average time to authenticate depending on the number of simultaneous connections we open against Kestrel:

Connections | Time
-------------- | ------
1 | 56 ms
2 | 65 ms
4 | 66 ms
8 | 76 ms
10 | 132 ms
12 | ~1 **s**
16 | ~4.5 **s**
32 | ~10 **s** _not all tasks complete_
64 | ~10 **s** _not all tasks complete_
128 | ~10 **s** _not all tasks complete_
256 | ~10 **s** _not all tasks complete_

I wrote a small program to measure this more isolatedly: https://gist.github.com/cesarbs/4409b712351adf621aae8420e1ea33de

On CoreFX 4.3 the call to `AuthenticateAsServerAsync` takes about 38 ms on average. On 4.4 that average is 99 ms (a discrepancy with Kestrel that I can't explain). Another interesting thing about this is that I only see those figures in the first request to the program - subsequent requests take around 3 ms to authenticate on both CoreFX versions.

cc @mikeharder

**Comment by davidsh** [comment]
What OS are you seeing this performance issue?

**Comment by karelz** [comment]
DO you have perf traces for the issue (ideally also for "working baseline"). That would speed up the investigation significantly.

**Comment by cesarbs** [comment]
@davidsh Server is running on Windows, client on Linux.

**Comment by cesarbs** [comment]
We're not seeing this issue anymore. It appears to have been fixed between versions 4.4.0-beta-24903-02 and 4.4.0-beta-24909-02.

While investigating this I thought this might be somehow related to the issue fixed by https://github.com/dotnet/corefx/pull/14852, but I did build corefx from source with that fix at the time and could still see the perf degradation in `AuthenticateAsServerAsync`.

Is there a changelog somewhere? We'd like to understand what could have been causing this.

**Comment by karelz** [comment]
Do you mean git history?

**Comment by cesarbs** [comment]
I hadn't realized you guys have commit messages for each version update. That's great, I'll try and find what change(s) might have affected this.

**Comment by benaadams** [comment]
coreclr had AVX2 switched off between beta-24815-03 https://github.com/dotnet/coreclr/pull/8624 and   beta-24904-02 https://github.com/dotnet/coreclr/pull/8788 - though don't think that explains why its only the first call; so probably not that...

**Comment by cesarbs** [comment]
This was fixed by the same change that fixed `WriteAsync` for us - https://github.com/dotnet/corefx/commit/5ccb1e47a8fe954c67026a577af99cb51166e68b

At the time I filed this issue I did build corefx with @stephentoub's fix to WriteAsync and saw the authentication performance issue, but I must have pointed to the wrong package instead of the one I had just built, or something of that sort.

Sorry for the false alarm.

**Comment by stephentoub** [comment]
Thanks, @cesarbs.  Just for future reference, do you know when the issue started?  Was it when Begin/EndXx were added back to Stream, or something else?

**Comment by cesarbs** [comment]
@stephentoub I don't know fore sure. @mikeharder, do we know the date when we started seeing this?

**Comment by mikeharder** [comment]
It's difficult to trace everything back to the CoreFx versions.  Our dev branch was fine on 12/21/16, then we first noticed this problem on 1/5/17, and it was fixed on 1/11/17.  Our automation was not working over the holidays, so I don't know when this started failing between 12/21/16 and 1/5/17.  Even if we did know exactly when it started failing in our dev branch, we'd then need to trace back the CoreFx version used at the time.
